sdks/python/apache_beam/io/sdf_line_source.py

```diff
@@ -0,0 +1,89 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import
+
+import os
+
+from apache_beam import Create
+from apache_beam import pvalue
+from apache_beam.io.restriction_trackers import OffsetRange
+from apache_beam.io.restriction_trackers import OffsetRestrictionTracker
+from apache_beam.transforms import core
+from apache_beam.transforms import ptransform
+
+
+class SdfReadLineRestrictionProvider(core.RestrictionProvider):
+  """ A `core.RestrictionProvider` implementation which provides
+  `OffsetRestrictionTracker`."""
+  def _get_file_size(self, file_name):
+    with open(file_name, 'rb') as f:
+      f.seek(0, os.SEEK_END)
+      return f.tell()
+
+  def initial_restriction(self, element):
+    return OffsetRange(0, self._get_file_size(element['file_pattern']))
+
+  def create_tracker(self, restriction):
+    return OffsetRestrictionTracker(restriction)
+
+  def split(self, element, restriction):
+    return restriction.split(element['desired_chunk_size'])
+
+  def restriction_size(self, element, restriction):
+    return restriction.size()
+
+
+class SdfReadLine(core.DoFn):
+  """ An SplittableDoFn which reads lines from file and as output."""
+  def process(
+      self,
+      element,
+      restriction_tracker=core.DoFn.RestrictionParam(
+          SdfReadLineRestrictionProvider())):
+    with open(element['file_pattern'], 'rb') as f:
+      start = restriction_tracker.start_position()
+      f.seek(start)
+      if start > 0:
```

Comment:
This probably deserves some comments. (I assume you're trying to see if we're starting at a line? What about multi-char newlines (e.g. \r\f)?